### PR TITLE
Support fused-expert MoE architectures (Qwen3.5-MoE)

### PR DESCRIPTION
**Pull request by maci0**
## The problem

`Model.get_layer_modules()` discovers per-layer residual-writing modules by iterating experts:

```python
with suppress(Exception):
    for expert in layer.mlp.experts:            # assumes an iterable of Linear experts
        try_add("mlp.down_proj", expert.down_proj)
```

This assumes each expert is a `Module` with a `.down_proj` `nn.Linear`. Several modern MoE
implementations instead **fuse all experts into batched tensors**. In transformers'
`Qwen3_5MoeExperts` (Qwen3.5-MoE) the block is:

```
Qwen3_5MoeSparseMoeBlock(gate, experts=Qwen3_5MoeExperts, shared_expert, shared_expert_gate)
Qwen3_5MoeExperts:
    gate_up_proj : Parameter [num_experts, 2*inter, hidden]
    down_proj    : Parameter [num_experts, hidden, inter]   # what writes to the residual
    # forward: nn.functional.linear(x, self.down_proj[expert_idx])
```

`for expert in layer.mlp.experts` yields nothing usable (it is not a `ModuleList`), so the
`suppress(Exception)` swallows it and **every routed expert is silently skipped**. Only
`attn.o_proj` and the shared-expert `down_proj` get abliterated. Because most of a MoE's refusal
behaviour lives in the routed experts, the optimizer cannot lower refusals no matter how many
trials it runs.

### Observed impact (a 256-expert, 40-layer Qwen3.5-MoE)
- Current Heretic: baseline 99/100, **best over 200 trials = 88/100** (nothing dips below 88).
- Orthogonalizing the fused `experts.down_proj` too (this PR), same directions and math: refusals
  drop to **0** on a held-out harmful set. The 88 floor was entirely the untouched experts
  (10240 expert output slices, 256 x 40, left bf16-identical).

## The fix

For a fused `down_proj` of shape `[E, hidden, inter]` and unit direction `v` in hidden space,
apply the projection removal Heretic already uses, `W_e <- W_e - weight * v (v^T W_e)`, vectorized
over experts:

```python
proj  = torch.einsum("h,ehi->ei", v, down_proj)        # [E, inter]
delta = weight * v.view(1, -1, 1) * proj.unsqueeze(1)   # [E, hidden, inter]
down_proj -= delta
```

Same per-layer direction and `max_weight`/`min_weight` schedule as the dense path.
`gate_up_proj` writes to the intermediate space (not the residual) and is left untouched, matching
the existing `up_proj`/`gate_proj` policy.

### Details
- The fused tensor is a `Parameter`, not an `nn.Linear`, so it cannot be LoRA-wrapped. It is edited
  **directly and reversibly**: the original is cached once and restored before each trial, so trials
  stay independent. `get_merged_model()` is unchanged because the edit is on the base weights.
- `reset_model()` restores the cached originals on the fast path and invalidates the cache on full
  reload (the params are recreated).
- Detection is conservative: only a 3-D `experts.down_proj` `Parameter` whose `shape[1] == hidden`
  is treated as fused. Iterable-expert MoEs (Mixtral, Phi-3.5-MoE) go through the existing path
  unchanged.

## Validation

End-to-end validation was done with an equivalent standalone implementation (per-layer directions,
the same fused orthogonalization) because a full Heretic run on the 35B needs a lot of memory; it
reproduced the 99 -> 0 result above. The integrated version in this PR applies the same math inside
`abliterate()` and reuses the existing direction/weight machinery. The repo already ships a
`tests/qwen3.5-moe` case, whose expected hashes will need regenerating since abliteration now
(correctly) touches the experts.

Happy to adjust naming/structure or add a unit test for the fused branch if you'd like.

### CODE REVIEW

**Comment by maci0** [comment]
Thanks for the review. Addressed in d9d6688:

- **Type annotations**: `_abliterate_fused_experts` is now fully annotated (`-> None`), and `_fused_experts_cache` is declared on the class body as `dict[int, tuple[torch.nn.Parameter, Tensor]]`.
- **Robustness**: the experts container is now located across `mlp` / `block_sparse_moe` / `feed_forward` / `moe`, and the fused param across `down_proj` / `w2` / `output_linear`, mirroring `get_layer_modules`. The `isinstance(fused, nn.Parameter) and dim() == 3 and shape[1] == hidden` guard keeps this a no-op for iterable-expert MoEs (Mixtral / Phi-3.5), which still go through the existing path.
- **Style**: `original_f` -> `original_fp32`, comments capitalized and punctuated.
- **Efficiency**: dropped the redundant `.to(fused.device)` on the cached original (it is cloned from the parameter, so already on device).

On testing: end-to-end validation was via an equivalent standalone implementation (the 35B needs a lot of memory for a full run), which took a 256-expert Qwen3.5-MoE from a 99/100 refusal baseline / 88/100 best-over-200-trials down to 0 on a held-out harmful set. Happy to add a unit test for the fused branch (synthetic experts tensor, assert `v` is removed from each slice and that restore-from-cache is exact) if that fits the suite.

**Comment by p-e-w** [comment]
This issue and its pros and cons have been discussed several times before, see e.g. #342.

> refusals drop to 0 on a held-out harmful set

What does "held-out" mean?

**Comment by maci0** [comment]
On "held-out": it just means those prompts were not part of the direction-fitting set. The direction came from `mlabonne/harmful_behaviors` (256) and `mlabonne/harmless_alpaca` (256); I then measured on 8 hand-written harmful instructions (lock-picking, malware, etc.) disjoint from that. But I should be straight: that is a small informal check (greedy decode, refusal-marker matching), not a refusals-at-KL Pareto on n=100. It shows the experts were the blind spot (99 -> 0 where attention-only cannot move off ~88), but it is not the rigorous comparison your bar calls for, and calling it a benchmark was an overstatement.

Having now read #342: that thread already covers this properly. rocker-zhang's `target_parameters` approach is generic across Qwen3.5 and GPT-OSS on one code path, measured on the Pareto front, ARA-aware, with a CPU test harness. This PR is a cruder subset: direct-weight, a separate method, Qwen3.5-only, no ARA story, no Pareto or GPT-OSS. It does not meet the requirements you laid out there (improvement over ARA, minimal model-specific code, one path for multiple fused-MoE families).

So I am closing this as superseded by the #342 direction rather than leave a weaker, redundant version open. If it is useful as independent confirmation: same `for expert in layer.mlp.experts` blind spot, and orthogonalizing the fused `down_proj` takes a 256-expert Qwen3.5-MoE from an ~88/100 floor to compliance. Thanks for the pointers.
